rulewerk-vlog/src/main/java/org/semanticweb/rulewerk/reasoner/vlog/VLogFastQueryResultIterator.java

```diff
@@ -0,0 +1,234 @@
+//			return size() >= this.maxCapacity;
+//		}
+//	}
+
+	/**
+	 * Simple cache for finding terms for VLog ids that is optimised for the case
+	 * where ids are inserted in a mostly ordered fashion. An LRU strategy is highly
+	 * ineffective for this as soon as the cache capacity is smaller than the number
+	 * of repeatedly used terms, since the cache entries there are always pushed out
+	 * before being needed again. This implementation will at least cache a maximal
+	 * initial fragment in such cases. It is also faster to write and requires less
+	 * memory.
+	 * 
+	 * @author Markus Kroetzsch
+	 *
+	 */
+	static class OrderedTermCache {
+		final private HashMap<Long, Term> terms = new HashMap<>();
+		final int maxCapacity;
+		private long maxId = -1;
+
+		public OrderedTermCache(int capacity) {
+			this.maxCapacity = capacity;
+		}
+
+		public Term get(long id) {
+			if (id > maxId) {
+				return null;
+			} else {
+				return terms.get(id);
+			}
+		}
+
+		public void put(long id, Term term) {
+			if (terms.size() < maxCapacity) {
+				terms.put(id, term);
+				if (id > maxId) {
+					maxId = id;
+				}
+			}
+		}
+	}
+
+	/**
+	 * The internal result iterator of VLog, returning numeric ids only.
+	 */
+	private final karmaresearch.vlog.QueryResultIterator vLogQueryResultIterator;
+	/**
+	 * The VLog instance. Used for resolving numeric ids to term names.
+	 */
+	private final VLog vLog;
+	/**
+	 * VLog ids of the previous tuple, with the last id omitted (since it is not
+	 * useful in caching).
+	 */
+	private long[] prevIds = null;
+	/**
+	 * RuleWerk terms corresponding to the previously fetched tuple, with the last
+	 * term omitted.
+	 */
+	private Term[] prevTerms = null;
+	/**
+	 * True if this is the first result that is returned.
+	 */
+	boolean firstResult = true;
+	/**
+	 * Size of the tuples returned in this result.
+	 */
+	int resultSize = -1;
+	/**
+	 * Cache mapping ids to terms.
+	 */
+	// final SimpleLruMap termCache;
+	final OrderedTermCache termCache;
+
+	private final Correctness correctness;
+
+	/**
+	 * Create a new {@link VLogFastQueryResultIterator}.
+	 * 
+	 * @param queryResultIterator
+	 * @param materialisationState
+	 * @param vLog
+	 */
+	public VLogFastQueryResultIterator(final karmaresearch.vlog.QueryResultIterator queryResultIterator,
+			final Correctness materialisationState, final VLog vLog) {
+		this.vLogQueryResultIterator = queryResultIterator;
+		this.correctness = materialisationState;
+		this.vLog = vLog;
+		// this.termCache = new SimpleLruMap(256, 64000);
+		this.termCache = new OrderedTermCache(130000);
+	}
+
+	@Override
+	public boolean hasNext() {
+		return this.vLogQueryResultIterator.hasNext();
+	}
+
+	@Override
+	public QueryResult next() {
+		final Term[] terms;
+		long[] idTuple = vLogQueryResultIterator.next();
+		terms = new Term[idTuple.length];
+
+		if (firstResult) {
+			resultSize = terms.length;
+			prevIds = new long[resultSize - 1];
+			prevTerms = new Term[resultSize - 1];
```

Comment:
Consider using `lastIndex = terms.length - 1` instead, which makes the following code much more readable.

Reply:
Can you explain? The last index of the prev* arrays is actually resultSize-2, so that name would seem confusing (but maybe I misread the comment).

Reply:
I was thinking about the last index of `terms`, but then `lastTerm` might be more specific and less confusing.

Reply:
Using an index (ordinal) as a size (quantity) does not seem so clear to me. After sleeping over it, I don't think this code needs any rewriting to be of acceptable quality ;-).

Reply:
Feel free to use a more appropriate name, but the point I object to is that you're using `resultSize - 1` in multiple places, specifically in `i < resultSize - 1` comparisons that are essentially off-by-one errors waiting to happen, so that should be the value stored in a variable.